src/codegen/estree/traverse.ts

```diff
@@ -0,0 +1,1524 @@
+import { TSESTree as ts } from '@typescript-eslint/types'
+import { NodeType } from './nodes'
+import { exhaustive } from '@/utils/typescript'
+
+type Visitor = {
+  [P in ts.Node['type']]?: (node: Extract<ts.Node, { type: P }>) => void
+} & {
+  enter?: (node: ts.Node) => void
+  leave?: (node: ts.Node) => void
+}
+
+export function traverse(node: ts.Node, visitor: Visitor) {
```

Comment:
This was a pain to write. It visits every possible type of node in typescript. All 168 of them.